src/v/storage/mvlog/versioned_log.cc

```diff
@@ -0,0 +1,171 @@
+readonly_segment::readonly_segment(std::unique_ptr<active_segment> active_seg)
+  : segment_file(std::move(active_seg->segment_file))
+  , readable_seg(std::move(active_seg->readable_seg))
+  , id(active_seg->id)
+  , offsets(model::bounded_offset_interval::checked(
+      active_seg->base_offset, model::prev_offset(active_seg->next_offset))) {}
+
+versioned_log::versioned_log(storage::ntp_config cfg)
+  : ntp_cfg_(std::move(cfg)) {}
+
+ss::future<> versioned_log::roll_unlocked() {
+    vassert(
+      !active_segment_lock_.ready(),
+      "roll_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ != nullptr, "Expected an active segment");
+    vlog(
+      log.info,
+      "Rolling segment file {}",
+      active_seg_->segment_file->filepath().c_str());
+    co_await active_seg_->segment_file->flush();
+    auto ro_seg = std::make_unique<readonly_segment>(std::move(active_seg_));
+    segs_.emplace_back(std::move(ro_seg));
+}
+
+ss::future<> versioned_log::create_unlocked(model::offset base) {
+    vassert(
+      !active_segment_lock_.ready(),
+      "create_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ == nullptr, "Expected no active segment");
+    auto new_path = fmt::format(
+      "{}/{}.log", ntp_cfg_.base_directory(), next_segment_id_());
+    vlog(log.info, "Creating new segment file {}", new_path);
+    auto f = co_await file_mgr_.create_file(std::filesystem::path{new_path});
+    auto active_seg = std::make_unique<active_segment>(
+      std::move(f), base, next_segment_id_, compute_max_segment_size());
+    active_seg_ = std::move(active_seg);
+    ++next_segment_id_;
+}
+
+ss::future<> versioned_log::apply_segment_ms() {
+    auto lock = active_segment_lock_.try_get_units();
+    if (!lock.has_value()) {
+        lock = co_await active_segment_lock_.get_units();
+    }
+    if (active_seg_ == nullptr) {
+        co_return;
+    }
+    const auto target_roll_deadline = compute_roll_deadline();
+    if (!target_roll_deadline.has_value()) {
+        co_return;
+    }
+    const auto now = ss::lowres_clock::now();
+    if (now >= target_roll_deadline.value()) {
+        vlog(
+          log.debug,
+          "Starting to roll {}, now vs deadline: {} vs {}",
+          ss::sstring(active_seg_->segment_file->filepath()),
+          now.time_since_epoch(),
+          target_roll_deadline->time_since_epoch());
+        co_await roll_unlocked();
+    }
+}
+
+ss::future<> versioned_log::close() {
+    // TODO(awong): handle concurrency. Should also prevent further appends,
+    // readers, etc.
+    if (active_seg_) {
+        co_await active_seg_->segment_file->close();
+    }
+    for (auto& seg : segs_) {
+        co_await seg->segment_file->close();
+    }
+}
+
+ss::future<> versioned_log::append(model::record_batch b) {
+    auto lock = active_segment_lock_.try_get_units();
+    if (!lock.has_value()) {
+        lock = co_await active_segment_lock_.get_units();
```

Comment:
`semaphore.wait` first checks if the units are available and if they are it returns a ready future. `get_units` call eventually calls this method so I don't think it make a lot of difference to use `try_get_units` here.

Reply:
Fair, this was a bit of premature "optimizing" in that we've done something similar to avoid entering coroutine machinery elsewhere, but I'm actually unsure how important it is given a lot of the lower levels probably return ready futures anyway.

Swapped out with get_units()